### ISSUE TITLE
Consider concept of phased pipelines

### ISSUE THREAD

**Issue by daveaglick** [statiqdev/Statiq#695]
A pattern has emerged where we often separate the data portion of a pipeline from the rendering of content. That's because to render things like sidebars or archives, we often need to know all of the data first. It's worth considering if this pattern should be explicitly supported by allowing pipelines to specify different phases. For example (not sure if these should be the actual phases or not, or how many phases seem reasonable):
* `Read`
* `Process`
* `Render`
* `Write`

All pipelines would execute the same phase, and then all pipelines would execute the next phase. After a phase, the pipeline would contain the resulting document from that phase, to be replaced by the resulting documents from the next phase.

For pipelines that don't specify phases, everything would happen in the `Process` phase (which also nicely handles backwards compatibility).

For the `Read` and `Write` phases, pipeline constructor overloads could be provided that implicitly set up `ReadFiles` and `WriteFiles` modules (we don't want to totally eliminate those modules though because a more complicated pipeline might use them in a single phase). We could even consider making the `Write` phase implicit: combined with the new `IDocument.Destination` property in #678, pipelines could automatically get a default `Write` phase that writes out any documents with a destination without having to do anything. The implicit `Write` phase shouldn't be added when only a `Work` phase is defined for backwards compatibility, and an option to opt-out of the implicit `Write` phase under any circumstance should also be provided.

Taking all that together, a pipeline might be defined like:

```
Pipelines.Add(
    "Pages",
    // Read phase, as specified by a globbing pattern
    "**/*.md",
    // Process phase    
    new IModule[]
    {
        new FrontMatter(new Yaml()),
        new Destination(".html")
    },
    // Render phase
    new IModule[]
    {
        new Markdown(),
        new Razor()
    });
    // Implicit Write phase using Destination property
```

Or maybe with a fluent API (I think I like this better):

```
Pipelines.Add("Pages")
    .Read("**/*.md")
    .Process(
        new FrontMatter(new Yaml()),
        new Destination(".html"))
    .Render(
        new Markdown(),
        new Razor());
    // Implicit Write phase using Destination property
```

Some additional thoughts on the phases:
* The `Read` phase is special because we know it won't depend on any other pipelines or documents and therefore we could execute all the `Read` phases from all pipelines in parallel.
* What happens if a pipeline needs to read files based on the `Process` result of another pipeline. I.e., in pipeline A I read some Markdown files in the `Read` phase and they contain file paths in their front matter that needs to be read in pipeline B but which won't be evaluated until pipeline A's `Process` phase? One answer is that the `ReadFiles` module for pipeline B just needs to go in it's `Process` phase instead of `Read`, and that's okay.
* The `Write` phase is also special like `Read` because it also shouldn't depend on any other pipelines or documents. Just like `Read` it could be executed in parallel with other pipelines. In addition, it could start executing immediately following a pipeline's `Render` phase while execution also moves to the next pipeline. The results of `Write` would then be cached and deferred until the global `Write` phase at which point pipelines who've already completed their `Write` phase would just swap in the result documents and others would block until theirs is done.
* If we do make assumptions about `Read` and `Write` not pulling information from other pipelines, modules and execution context methods that get documents from other pipelines should throw when in one of these phases.

**Comment by LokiMidgard** [comment]
Maybe the phases could be compleatly custom and easyly defined by the user.

So the user provide a set of phases in the beginning, and evry pipline will run those phases.

**Comment by daveaglick** [comment]
@LokiMidgard That's an excellent thought, and was basically how I've been thinking about this for weeks...up until I started writing the issue for #694 (which then caused me to finally write this idea out too). By having the phases predefined, we can use them to make smarter decisions around things like performance. It enforces a pattern, but each phase also has a particular intent that we can leverage for other features.

**Comment by daveaglick** [comment]
Been working on this for v3 and settled on the following approach:

- Four phases per pipeline: `Read`, `Process`, `Render`, `Write`
- Each phase doesn't need to contain modules for each pipeline
- Pipelines can declare dependencies which affect the `Process` phase (see below)
- Pipelines and phases are run in parallel as defined below (#486, #189)
- Only the output of the `Process` phase is made available to other pipelines
- A new `IDocument.GetHash()` method will be added to supply hashes for caching (see the `Process` phase below, #694)
- A pipeline can be marked as "isolated" in which case it isn't subject to dependency checks between pipelines or phases
  - Isolated pipelines will throw if modules access the global documents collection (`IExecutionContext.Documents`)
  - Isolated pipelines don't add output documents from their `Process` phase to the global documents collection (though we still have to track the outputs for hash-based caching purposes)
  - The difference between an isolated pipeline and one that doesn’t have any dependencies or dependents is that the latter will still wait until all `Process` phases are complete before starting `Render` and can access the global documents collection.
  - Useful for pipelines like Sass processing that generally should just go through modules in parallel as quickly as possible
- The location metadata like `RelativeFilePath` will be removed in favor of a single `IDocument.Destination` property to make implicit `Write` phases easier (#678)
- The global documents collection will be tweaked and will no longer contain documents for each pipeline after every module - the output documents will only be added once the `Process` phase is complete
  - Accessing from an isolated pipeline will throw
  - Accessing documents from a pipeline in `Process` for which there’s no dependency will throw
  - Accessing from the `Read` or `Write` phase will throw

Phases:
- `Read`
  - All `Read` phases execute immediately and in parallel
  - Intended only to grab files from disk or some other source
- `Process`
  - Follows the `Read` phase _and_ `Process` phases from other dependent pipelines (a circular dependency check will need to be performed during engine initialization)
    - Unless the pipeline is isolated
  - On subsequent executions (I.e., when input files change) if _all_ document hashes from input documents as a result of the `Read` phase _and all_ hashes from output documents from the `Process` phase of dependent pipelines match (and the counts of each match), the `Process` phase will be skipped and cached output documents from the previous `Process` run-through will be output
    - This behavior can be opted out using a flag
- `Render`
  - `Render` phases don't execute until _all_ `Process` phases have completed
    - Unless the pipeline is isolated
  - All `Render` phases execute in parallel and can access documents from the `Process` phase of any pipeline
  - The `Render` phase isn't part of hash-based output document caching since rendering often uses external files like layouts that aren't represented by documents (see https://github.com/Wyamio/Wyam/issues/694#issuecomment-412168009)
- `Write`
  - Immediately follows the `Render` phase and executed in parallel

![2019-04-26_11h48_59](https://user-images.githubusercontent.com/1020407/56820219-4e619580-6819-11e9-9643-b191c137c300.png)

This creates some pipeline registration challenges. We want adding pipelines to be easy with simple cases while exposing the full capabilities for more complex ones:
- The existing `AddPipeline()` methods that take a single sequence of modules will continue to be available
  - The modules will be added to the `Process` pipeline
  - A dependency will be added to the pipeline defined immediately before this one
  - The implicit `Write` phase won't be added (I.e., just the `Process` phase will contain modules)
  - This has the effect of acting just like it currently does
- New methods that use a delegate/builder pattern will be added to allow adding more complex pipelines
  - Overloads that allow specifying a globbing pattern(s) will automatically populate the `Read` phase with the appropriate `ReadFiles` module
  - An implicit `WriteFiles` module will be added to the `Write` phase (unless otherwise specified with a flag/`NoWrite()` method or similar)
    - Can be fine-tuned by specifying a delegate/filename/etc. and appropriate `WriteFiles` module will be created

Since this issue now touches a bunch of other ones, I'll track the overall progress of these related changes here and close the other issues as those parts are completed.

**Comment by daveaglick** [comment]
Implemented in vnext branch
